New features: chunking and compression

khou2020 opened the pull request:
This PR is to add data chunking and compression features. By using additional metadata and manipulate the space between data object file offsets, these features can be developed without breaking the NetCDF file format specifications.

More information about the design and implementation are described in K. Hou, Q. Kang, S. Lee, A. Agrawal, A. Choudhary, W. Liao. [Supporting Data Compression in PnetCDF](https://ieeexplore.ieee.org/document/9671998), published in the International Conference on Big Data, 2021.

An example program is available in [./examples/C/chunk_compress.c](https://github.com/Parallel-NetCDF/PnetCDF/blob/chunk/examples/C/chunk_compress.c)

khou2020 committed:
add ncmpi_var_get/set_chunk and ncmpi_var_get/set_filter

khou2020 committed:
add config option to enable chunking feature

khou2020 committed:
add chunk drivere

khou2020 committed:
enable chunk driver when chunking hint is set

khou2020 committed:
rename default filter hint to nc_chunk_default_filter

khou2020 committed:
add documentation about chunked I/O driver

khou2020 committed:
bugfix:chunk driver build condition is ENABLE_CHUNKING

khou2020 committed:
always build chunk related APIs

khou2020 committed:
bug fix: add ncchunkio to DIST_DIR

khou2020 committed:
bugfix: add new m4 scripts to extra dist

khou2020 committed:
bugfix: add ncchkioi_profile_timers.m4 to extra_dist

khou2020 committed:
bugfix: add nchunk io to m4 include

wkliao committed:
pncio, a new I/O driver

wkliao committed:
define ENABLE_COMPRESSION if --enable-zlib or --enable-sz is set

wkliao committed:
add example of using chunking and compression

wkliao committed:
run chunk_compress.c only when compression is enabled

wkliao committed:
Remove unused variable, initialize err to NC_NOERR

wkliao committed:
add chunking example, examples/C/chunk_io.c

wkliao committed:
Add verbose to print which processes perform compression/decompression

wkliao committed:
comment out unused printf

wkliao committed:
test: add examples/C/chunk_2D.c which uses 2D checkerboard partitioning pattern

wkliao committed:
examples/C/chunk_2D.c one wait_all for all nonblocking reads

wkliao committed:
chunk_2D.c: fix use of NX by mistake

wkliao committed:
use MPI_COMM_SELF to call MPI_Pack and MPI_Unpack

wkliao committed:
bug fix: sbuf index, array of size nsend

wkliao commented:
Since you mentioned HDF5, I wonder if there is any thing missing in HDF5 but is in this PR of PnetCDF.

wkliao committed:
chunk: fix printing expected number of records

wkliao committed:
chunk_2D.c add partitioning along time for decompression

wkliao commented:
For such unusually installation, I suggest to set CPPFLAGS environment variable
to the SZ include path when running configure. Please let me know if that works.

dqwu commented:
> For such unusually installation, I suggest to set CPPFLAGS environment variable to the SZ include path when running configure. Please let me know if that works.

Actually that is the default installation of SZ.

Setting CPPFLAGS environment variable to the SZ include path before calling configure works:
`export CPPFLAGS="-I /sz/install/path/include/sz"`

wkliao committed:
bug fix: all ranks participate wait call whenin collective data mode

wkliao committed:
add test program for chunking nonblocking API call

wkliao commented:
Hi, @dqwu 

Thanks. I was able to reproduce the error.
I pushed a fix. Please let me know if it fixes the problem.

wkliao committed:
update RELEASE_NOTES for version 1.15.0-alpha

wkliao committed:
add note and URL of 1.15.0-alpha release

wkliao committed:
new error core NC_EMULTIDEFINE_HINTS for inconsistent I/O hints set by users

wkliao committed:
rename hint variables used internally

cb_read     -> romio_cb_read
cb_write    -> romio_cb_write
ds_read     -> romio_ds_read
ds_write    -> romio_ds_write
no_indep_rw -> romio_no_indep_rw

wkliao committed:
One Bcast and Allreduce to check consistency of all I/O hints

wkliao committed:
move setting is_open inside of open call

wkliao committed:
parallel_run.sh run ncmpidiff based on driver mode

wkliao committed:
Bug fix: check write buffer size consumed after increment

wkliao committed:
make ncmpi_close() returns the same error code

wkliao committed:
duplicate MPI info for non-INA aggregators in case later switch to indep mode

wkliao committed:
pncio_write_str_naive.c assert disp and offset are 0s

wkliao committed:
benchmarks/C/aggregation.c add indep I/O mode option

wkliao committed:
benchmarks/C/pnetcdf_put_vara.c add indep I/O mode option

wkliao committed:
benchmarks/C/write_block_read_column.c add indep I/O mode option

wkliao committed:
benchmarks/C make ptest problem size larger

wkliao committed:
FLASH-IO/parallel_run.sh separate cases for running ncmpidiff

wkliao committed:
test/C add an option to run in indep I/O mode

wkliao committed:
test/nonblocking/bput_varn.m4 add an option to put multiple variables

wkliao committed:
column_wise.m4 add testing for indep I/O mode

wkliao committed:
flexible_bput.c add testing for indep I/O mode

wkliao committed:
testcases/alignment_test.c increases problem size

wkliao committed:
tst_multi_redefine.c tests entering define mode multiple times

wkliao committed:
PNETCDF_PROFILING: print profiling timings by world rank 0 only

wkliao committed:
Github actions: disable profiling and add MIMIC_LUSTRE

wkliao committed:
FLASH-IO/parallel_run.sh test independent data mode

wkliao committed:
alignment_test.c test indep data mode before collective

wkliao committed:
buftype_free.c test independent data mode and check read buffer contents

wkliao committed:
tst_multi_redefine.c checks variable contents after growing file header

wkliao committed:
wrf_io.c add option of using blocking APIs

wkliao committed:
wrf_io.c add a check of file contents after write when in debug mode

wkliao committed:
testcases/flexible.c add test of independent data mode

wkliao committed:
buftype_free.c define INDEP_MODE and COLL_MODE for better readability

wkliao committed:
flexible.c increase I/O tests size

wkliao committed:
flexible2.c add testing of independent data mode

wkliao committed:
ncmpidiff.c print command-line options when first countering a difference

wkliao committed:
bug fix: buf_view.off[] is constructed based on the user buffer, buf

wkliao committed:
print an extra INA profile timer

wkliao committed:
ncmpio_intra_node.c copying noncontiguous write buffer to contiguous is much faster

wkliao committed:
testcases/varn_int.c fix varid used in testing

sbelsk committed:
ncadios: Fix typo causing compilation error

wkliao committed:
testcase/varn_int.c add tests independent data mode and set nc_ibuf_size to 0

wkliao committed:
testcase/flexible_var.c increase problem size and test set nc_ibuf_size to 0

wkliao committed:
testcase/flexible_varm.c add tests independent data mode and set nc_ibuf_size to 0

wkliao committed:
testcase/iput_all_kinds.m4 add tests independent data mode and set nc_ibuf_size to 0

wkliao committed:
bug fix: ensure str_factor <= overstriping_ratio * total_num_OSTs

wkliao committed:
pncio_lustre_open.c printf to stdout when in DEBUG mode

wkliao committed:
Bug fix: pncio_lustre_open.c ensure OST ID is within range of numOSTs

Skip calling llapi_layout_ost_index_set() if hint start_iodevice is not
set by the user

wkliao committed:
INA reuse recv_buf when possible to reduce memory footprint

wkliao committed:
INA report memory footprint high-water mark instead

wkliao committed:
Enable ncmpi_inq_header_size() to be called in define mode.

This API returns the file header size with metadata defined by the time
of the call. This information can be helpful to pick proper values for
arguments h_minfree, v_align, v_minfree, r_align when calling
ncmpi__enddef() to allocate a sufficiently large free space for file
header extent and variable data sections to grow without moving data
already stored in the file, i.e. when adding new variables, dimensions,
or attributes.

Test program, tst_inq_header_size.c tests this change.

wkliao committed:
valgrind warn: to avoid unnecessary increment of fd->flat_file.idx

wkliao committed:
ncmpio_getput.m4 check zero-sized request as buf pointer may not be initialized

wkliao committed:
Bug fix: move data after adding new variables

This handles the case when file header extent does not grow after adding
new variable. In addition, the case when adding new record variables
does not grow the starting offset of record variable section.

sbelsk committed:
Add CI job for ADIOS backend

sbelsk committed:
Fix syntax error in test script

wkliao committed:
testcases/file_create_open.c test with adding an MPI info object

wkliao committed:
tst_redefine.c update hint alignment calculation

wkliao committed:
RELEASE_NOTES rewording

wkliao committed:
Ensure hint nc_var_align_size supersedes nc_header_align_size

wkliao committed:
add fix_end, end offset of last fix-sized variable

wkliao committed:
Revisit of aligment hints for setting the offset of header and data section

The setting is based on the followings.

Each time ncmpi__enddef() is called, use env variable and MPI info
object to recalculate ncp->v_align and ncp->r_align, which are later
used to calculate the file starting offsets for fix-sized variable
section and record variable section.

    if (env variable of v_align is set)
        ncp->v_align = env value
    else if (MPI info hint of v_align is set)
        ncp->v_align = hint value
    else if (v_align passed to ncmpi__enddef() > 0)
        ncp->v_align = v_align + 4 - v_align % 4;
    else /* Not set by user, <= 0 */
        ncp->v_align = 0;

if has fix-sized variables
    if ncp->v_align > 0    /* v_align is set by users */
        use it
    else
        ncp->v_align = default 512

    * use ncp->v_align to set ncp->begin_var

    if ncp->r_align > 0    /* r_align is set by users */
        use it
    else
        ncp->r_align = default 4

    * use ncp->r_align to set ncp->begin_rec

else
    discard ncp->v_align
    discard ncp->v_minfree

    if ncp->r_align > 0     /* r_align is set by users */
        use it
    else
        ncp->r_align = default 512

    * use ncp->r_align to set ncp->begin_rec
    * set ncp->begin_var == ncp->begin_rec

Do not reuse ncp->v_align or ncp->r_align in next ncmpi_enddef() !

wkliao committed:
WRF-IO: fix run script when NetCDF4 is enabled

wkliao committed:
test/CXX programs require at least 2 command-line arguments

wkliao committed:
testcases/null_args.m4 skip calls to ncmpi_fill_var_rec when NetCDF4 is enabled

Because NetCDF4 does not support API ncmpi_inq_header_size

wkliao committed:
skip tst_grow_data.c when NetCDF4 is enabled

Because NetCDF4 does not support API ncmpi_inq_header_size

wkliao committed:
tst_inq_header_size.c skip tests for NetCDF4 files

Because NetCDF4 does not support API ncmpi_inq_header_size

wkliao committed:
fix warning: declaration of 'i' shadows a previous local [-Wshadow]

wkliao committed:
tst_data_move.c checks whether the data movement subroutines are called

When there is a sufficiently large free space in the header extent and
fix-sized variable section, such that all successive adding new
variables will not change the starting offsets of both fix-sized and
record variable sections.

wkliao committed:
tst_grow_data.c rename NY/NX to LON/LAT

wkliao committed:
Obtain the total number of OSTs, given a folder

This can be done by calling the Lustre user API.
int llapi_get_obd_count(char *dir, int *count, int is_mdt);

wkliao committed:
ncmpidiff.c - add stat checks for input files

wkliao committed:
Move all calls of getenv() to file create/open

wkliao committed:
ncoffsets.c add command-line option '-d' to replace getenv PNETCDF_VERBOSE_DEBUG_MODE

wkliao committed:
Add note on how environment variable PNETCDF_VERBOSE_DEBUG_MODE works

wkliao committed:
pass environment variables, env_mode, from dispatcher to all drivers

wkliao committed:
make safe mode a part of ncp->flags

The env_mode passed from dispatcher is now saved into ncp->flags.
This avoids costly calls to genenv().

wkliao committed:
Add functions to check env variable PNETCDF_RELAX_COORD_BOUND

wkliao committed:
Call is_relax_coord_bound() instead of check macro RELAX_COORD_BOUND

Whether RELAX_COORD_BOUND is enabled/disabled depends on
1. option --disable-relax-coord-bound is set at the configure time
2. env variable PNETCDF_RELAX_COORD_BOUND is set at the run time and
   its value being 0 or 1.
is_relax_coord_bound() checks the combinations of the above.

wkliao committed:
fix a missing printf argument

wkliao committed:
testcases/varn_int.c check NULL before free()

wkliao committed:
PNCIO_Calc_my_req() remove argument fd_start as it is not used

wkliao committed:
bug fix: special care for when fd->flat_file.count == 0, entire file is visible

* save fd->flat_file and restore it before leaving the collective I/O
  call, i.e. restore the original fileview.
* When fd->flat_file.count == 0, the entire file is visible and we set
  and change fd->flat_file.count, fd->flat_file.off[0], and
  fd->flat_file.len[0] which will be used in PNCIO_Calc_my_req() to
  construct my_req.
* Similary for Lustre collective write subroutines.

wkliao committed:
Try issend/irecv to replace MPI_Alltoall

wkliao committed:
When MIMIC_LUSTRE is not enabled, set hint cb_nodes=2 for all testings

wkliao committed:
Github action: use OpenMPI 5.0.8

wkliao committed:
fix bash script error of "test: =: unary operator expected"

wkliao committed:
align print line

wkliao committed:
add an option of MPI_Allreduce besides MPI_Allgather

However, MPI_Allreduce does not appear to be faster on Perlmutter at
NERSC.  Maybe it is on some machines.

wkliao committed:
Check file system type at rank 0 only and then bcast result

wkliao committed:
WRF-IO add a benchmark option of moving data section when header grows

wkliao committed:
ptest - add tests of romio_no_indep_rw = true and false

wkliao committed:
ncmpio_enddef.c - participate collective I/O even if zero-sized requests

Also do proper type casting MPI_Offset for C constant MOVE_UNIT

wkliao committed:
ncmpio_file_io.c returns immediately when file handler is NULL

wkliao committed:
Set buf_view.type = MPI_BYTE for zero-sized requests

wkliao committed:
check hint romio_no_indep_rw and set the flag accordingly

wkliao committed:
pncio_aggregate.c - add assert(fd->flat_file.count > 0);

wkliao committed:
Bug fix: collectively write number of records when in collective data mode

When hint no_indep_rw is set to true, all MPI-IO/PNCIO calls must be
collectively when in collective data mode. When in independent data
mode, NC_HCOLL should be ignored

wkliao committed:
ncmpio_fill.c fill buffer is always contiguous

wkliao committed:
Github actions: change timeout to 90 minutes

wkliao committed:
hints.c add print prefix 'info'

wkliao committed:
New hint nc_data_move_chunk_size

When adding new data objects into an existing file, the data section may
need to be moved to a higher file offset.  The movement is performed in
chunks. This hint allows users to customized the chunk size. The default
is 1048576 bytes, i.e. 1 MiB.

wkliao committed:
wrf_io.c print hint nc_data_move_chunk_size

wkliao committed:
test/parallel_run.sh - set nc_data_move_chunk_size to 100

In addition, define DRY_RUN and VERBOSE and 2 functions for debugging purpose.

wkliao committed:
Github actions: separate tests of MIMIC_LUSTRE

wkliao committed:
wrf_io.c print hint cb_node_list

wkliao committed:
move setting hint data_chunk to after file open

Value of data_chunk depends on file striping_unit
It performs best when set to striping_unit.

wkliao committed:
Improve data section movement

* use file striping unit size as moving chunk size by default.
* align write offsets with the striping units to coincide the
  Lustre collective write file domain alignment

wkliao committed:
parallel_run.sh setenv MIMIC_STRIPE_SIZE to 1048576

wkliao committed:
Github actions: increase timeout to 120 minutes

wkliao committed:
Separate ncmpidiff core subroutines into a file and build it as a library

This enables PnetCDF test programs to call the ncmpidiff core subroutine
to make comparison of two files.

wkliao committed:
test: define TESTOUTDIR in config.h

wkliao committed:
test/common: add C struc loop_opts and tst_main()

Enable all test programs to use the common codes for
* loops of various run-time options
* command-line argument processing

Call assert(0) for all errors. Using a core dump is easier for debugging.

williamjallen committed:
Pin GHA dependencies by hash

Hash-pinned dependencies are an important requirement enforced by the OpenSSF scorecard tool: https://github.com/ossf/scorecard/blob/main/docs/checks.md#pinned-dependencies

This commit pins all GitHub Actions workflows to a commit hash.

wkliao committed:
ncmpio_enddef.c -Wshorten-64-to-3 warning

wkliao committed:
test/testcases -Wshorten-64-to-3 warning

williamjallen committed:
Configure Dependabot updates for GitHub Actions

This commit adds a basic Dependabot configuration for GitHub Actions workflow updates. In addition to keeping dependencies fresh, adding a Dependabot configuration allows the project to receive Dependabot security alerts if configured in the project settings: https://docs.github.com/en/code-security/dependabot/dependabot-security-updates/configuring-dependabot-security-updates

wkliao committed:
ncmpidiff.c - ncmpidiff_core() return data type

* ncmpidiff_core() return data type of MPI_Offset (number of differences)
* ncmpidiff.c switch default to call usage()
* free allocated buffers

dependabot[bot] committed:
Bump actions/checkout from 4.3.1 to 6.0.1

Bumps [actions/checkout](https://github.com/actions/checkout) from 4.3.1 to 6.0.1.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/34e114876b0b11c390a56381ad16ebd13914f8d5...8e8c483db84b4bee98b60c0593521ed34d9990e8)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-version: 6.0.1
  dependency-type: direct:production
  update-type: version-update:semver-major
...

Signed-off-by: dependabot[bot] <support@github.com>

wkliao committed:
test/header make use of tst_main()

wkliao committed:
Github action: try /tmp and $RUNNER_TEMP

wkliao committed:
Bug fix: indivisible case when nprocs_my_node % naggrs_my_node > 0

calculate the number of ranks in each INA group, ncp->num_nonaggrs
and the local-node of each rank's aggregator's rank, first_rank.

wkliao committed:
test/C - merge pres_temp_4D_wr.c and pres_temp_4D_rd.c

The two programs can be combined with write first followed by read.
Also make use of tst_main()

wkliao committed:
test/cdf_format make use of tst_main()

wkliao committed:
test/cdl make use of tst_main()

wkliao committed:
test/testcases make use of tst_main()

wkliao committed:
m4/utils.m4 add CDF5_EXTRA_ITYPE_LIST and _all option

wkliao committed:
test/nonblocking make use of tst_main()

wkliao committed:
Github action: make -s LIBTOOLFLAGS=--silent V=1

wkliao committed:
test/F90 make use of get_args()

wkliao committed:
test/largefile make use of tst_main()

wkliao committed:
test/fandc Makefile.am remove use of variable TESTPROGRAMS

wkliao committed:
test/CXX update parallel_run.sh and seq_runs.sh

wkliao committed:
test/adios clean files all have tail name nc or bb

wkliao committed:
test/nc_test clean files all have tail name nc or bb

wkliao committed:
test/nc4 clean files have tail name nc or bb. remove unused variables

wkliao committed:
test/nf90_test clean files all have tail name nc or bb

wkliao committed:
test/nf_test clean files all have tail name nc or bb

wkliao committed:
test/subfile Makefile.am remove use of variable TESTPROGRAMS

wkliao committed:
test/burst_buffer - make use of tst_main()

clean files all have tail name nc or bb
Makefile.am - remove use of variable TESTPROGRAMS

wkliao committed:
dtype_decode.c warning uninitialized variable

wkliao committed:
ncoffsets.c use uint32_t and uint64_t in byte swap

When compiled with gcc -O2, incorrect values of byte-swap can occur

wkliao committed:
ncvalidator.c use uint32_t and uint64_t in byte swap

When compiled with gcc -O2, incorrect values of byte-swap can occur

wkliao committed:
benchmarks/C skip test no_indep_rw_hint and ptests on 4 and 6 processes

wkliao committed:
benchmarks/FLASH-IO skip test no_indep_rw_hint and ptests on 4 and 6 processes

wkliao committed:
benchmarks/WRF-IO skip test no_indep_rw_hint and ptests on 4 and 6 processes

wkliao committed:
examples make ptests run 4 and 6 processes

Remove testing hint romio_no_indep_rw

wkliao committed:
nc_test/test_write.m4 -Wstringop-truncation

wkliao committed:
testcases/put_all_kinds.m4/iput_all_kinds.m4 remove need of calling redef()

wkliao committed:
tst_grow_data.c use nonblocking put

wkliao committed:
bug fix: pncio - condition on info object duplication

wkliao committed:
free ncp->mpiinfo before getting a new one

wkliao committed:
ncbbio_log.c warning integer_sign_change

wkliao committed:
ncbbio_log.c use NULL in calls to realpath()

wkliao committed:
github actions permission fix, recommended by code scanning alerts

wkliao committed:
Github action: combinations of Ubuntu/MacOS, MPICH/OpenMPI, UFS/Lustre

wkliao committed:
Github actions: delete yaml files that have been moved to main.yml

wkliao committed:
Github actions: test NetCDF4 and ADIOS

Also merge ubuntu_openmpi_adios.yml into netcdf4_adios.yml

wkliao committed:
fix compilation warnings found by github code scanning

* test/testcases/modes.c
  - Time-of-check time-of-use filesystem race condition

* Others
  - Too few arguments to formatting function

wkliao committed:
Github actions: fix OpenMPI network issue when running make ptest

* Specify the loopback interface: Explicitly instruct Open MPI to use
  the local loopback interface (lo0) for communications by setting the
  OMPI_MCA_btl_tcp_if_include environment variable. In your terminal
  (e.g., zsh or bash), run the following command:
  + export OMPI_MCA_btl_tcp_if_include=lo0
  Then, try running your MPI program again.

* Set a short temporary directory path: In some cases, long file names in
  the default temporary directory can cause issues on macOS. Setting the
  TMPDIR environment variable to a shorter path like /tmp can help:
  + export TMPDIR=/tmp
  After setting this, try running your MPI program.

* Error message:
  [iad20-eo1205-ccaaba69-b68d-4eae-afa9-dee869e099ae-2690965F61F0.local:90667]
ptl_tool: problems getting address for index 0 (kernel index -1)
  --------------------------------------------------------------------------
  You requested support for tool connections, but no available
  FAIL: nprocs=./pnetcdf-permute ---- pnetcdf-permute
  interfaces were found after applying any include or exclude directives:

    Include:    NULL
    Exclude:    NULL
    Available:

  Please adjust your include or exclude to allow selection of an
  available interface.

wkliao committed:
update github CI status badge

williamjallen committed:
Change CI permissions to read-only

The `pull-requests: write` [permission](https://docs.github.com/en/actions/reference/workflows-and-actions/workflow-syntax#permissions) allows the GITHUB_TOKEN to be used to label, comment on, and perform other interactions with the pull request.  Neither of these workflow files require these permissions.  Minimum permission is best practice.

wkliao committed:
testutils.c: print timing result for each test when in verbose mode

wkliao committed:
UFS: obtain file system block size as striping size

wkliao committed:
tst_redefine.c - uses nonblocking APIs

dependabot[bot] committed:
Bump actions/checkout from 6.0.1 to 6.0.2

Bumps [actions/checkout](https://github.com/actions/checkout) from 6.0.1 to 6.0.2.
- [Release notes](https://github.com/actions/checkout/releases)
- [Changelog](https://github.com/actions/checkout/blob/main/CHANGELOG.md)
- [Commits](https://github.com/actions/checkout/compare/8e8c483db84b4bee98b60c0593521ed34d9990e8...de0fac2e4500dabe0009e67214ff5f5447ce83dd)

---
updated-dependencies:
- dependency-name: actions/checkout
  dependency-version: 6.0.2
  dependency-type: direct:production
  update-type: version-update:semver-patch
...

Signed-off-by: dependabot[bot] <support@github.com>

wkliao committed:
testutils.c fix logic of skip_diff

wkliao committed:
testcases/flexible.c return error immediately found

wkliao committed:
increase test array sizes

wkliao committed:
fix compilation warnings

wkliao committed:
add a new hint nc_striping

'nc_striping' -- When creating a new file on the Lustre file system, this
hint advises PnetCDF to set the file's striping configuration. The hint
value is either "auto" or "inherit". The former sets the striping count to
the number of compute nodes found in the MPI communicator passed to
`ncmpi_create()`. The latter makes the new file to inherit the folder's
striping settings if the folder's striping is set. This hint's default is
"auto". Hint 'nc_striping' is ignored when MPI-IO hints `striping_factor`
and `striping_unit`, are set.

wkliao committed:
spell check.

github copilot code quality
    [nitpick] Variable name `dir_name` uses underscore convention while
    `path_dup` uses underscore as well, but the naming could be more
    consistent with the project's naming conventions. Consider using
    consistent naming patterns throughout the file.

wkliao committed:
rename internal variable striping to nc_striping to be consistent

wkliao committed:
add node IDs of all processes as an attribute cached in MPI communicator

In case the same communicator is used to create/open multiple files, we
can cache the node IDs of all processes as attributes of the MPI
communicator, so they can be reused every time the same communicator is
used to create/open a file in the same application run. Using the cached
node IDs can avoid repeatedly calling MPI_Get_processor_name(),
MPI_Gather(), and MPI_Bcast() to construct the node IDs.

Note cb_nodes, the number of I/O aggregators, and rank IDs of I/O
aggregators cannot be reused by another file, because cb_nodes is a hint
that may be set to a different value each time file is created/opened.
Thus cb_nodes and rank IDs of I/O aggregators must be recalculated at
each file create/open.  Fortunately, Lustre_set_cb_node_list() and
GEN_set_cb_node_list() do not make any MPI communication calls. The cost of
recalculating cb_nodes and rank IDs is expected to be small.

wkliao committed:
receive node IDs of all processes from dispatcher's file create/open

wkliao committed:
Update data structure for storing node IDs and number of nodes

wkliao committed:
Update file create/open APIs for receive node IDs of all processes

Node IDs and the number of compute nodes are generated during file
creation or opening at the dispatcher.

wkliao committed:
new example program for chunking/compression feature: chunk_compress_FLDS.c

wkliao committed:
rebase master branch

* Note chunking/compression feature is only supported when using the
  MPI-IO driver. Hint nc_pncio must be set to "disable".
